### DISCUSSION TITLE
Restructure the settings page

### DISCUSSION THREAD

**Discussion by iqnite** [ScratchAddons/ScratchAddons#7712] [Ideas]
# The situation

By now, Scratch Addons has more than 170 addons. Most of those addons are either minor or only useful to more advanced users, yet they all take up an equal portion of the settings page. While this allows Scratch Addons to have a scalable and simple structure, it's hell for users who have just installed the extension to scroll through all those features to find those they installed the Extension for. The fact that the addons are ordered alphabetically (*ew*) instead of having a logical order (like Visual Studio Code) makes it even worse.

# Proposed solutions

Scratch Addons already has categories, but they're way too crowded. Many Issues (#5389 #5209 #5407 #6099 and more) proposed an onboarding tour, which shows popular addons to the user, making a friendlier first-time user experience. A "Related addons" feature (#6104) is in the works, and probably coming soon. However, all these implementations are closer to the concept of an **app store** rather than a **settings page**. This gives the advantage of an environment open to developers, with individual, compatible addons, where addons can be developed and integrated quickly, all the same way. From the development side, this is genius, but from the user experience side, it's overwhelming.

# A new approach

A settings page should provide a hierarchy, where common settings are at the surface, and advanced fine-tunings can be found if you dig deeper. Take the Windows settings: You have general categories on the side ("System", "Customization", "Security", ...). When you open one of those categories, you have a main page with quick options at the top (Theme picker in "Customization", Connection status in "Network", ...) and subcategories below them. These subcategories continue in a tree of more subcategories, where the more specific settings can be found deeper inside.
This concept eliminates the need for an onboarding tour, as the most popular settings usually are also the most general. A lot of those settings would be enabled by default, so the user doesn't even need to scroll through them, and can disable them if they don't like them. **It's hard to search for an addon to enable if you don't even know if it exists, but it's much easier to search for an addon to disable if you see it in action.**
There could also be a way to access all settings on one page (like in the current experience) by clicking on an "All settings" button.

# Development considerations

The Addon Manifest system would need to be reworked. The tags need to be hierarchical, so the settings page can assign them correctly. But quick integration is less important than back in the good old days when there was an update with 7 new addons every month. Today, Scratch Addons updates are much smaller and stabler, which gives more time to manually integrate into the new settings system.

# Suggested hierarchy

<details>
<summary>Click to open (text wall ahead!)</summary>

*Subject to change. Easter eggs and deprecated addons are not included.*

- *Editor*
  - *General*
    - File drag and drop
    - Sprite folders
    - Variable manager
  - *Code editor*
    - *Blocks*
      - Block switching
      - Jump to custom block definition
      - Save blocks as image
      - Copy reporter values
      - Snap to grid
      - Rename broadcasts
      - Rearrangeable custom block inputs
      - Extra key options
      - Point towards random direction block
      - Switch variables between "For all sprites" and "For this sprite only"
    - Developer tools
    - Name scripts before placing in backpack
    - *Editor comments*
      - Editor comment previews
      - Better editor comments
  - *Costume editor*
    - Onion skinning
    - Snapping
    - Skewing
    - HD image uploads
    - Hex color picker
    - Opacity slider
    - Bitmap images copying
  - *Sound editor*
    - Echo effect button
  - *Fixes* 
    - Do not shift pasted items
    - Do not automatically run duplicated blocks
    - Do not automatically space overlapping scripts
    - SVG uploads fix
    - Fix sprite pane glitching
    - Ctrl+Click to run scripts
    - Place backpack code at mouse
    - Non-draggable sprites in editor
- *Customization*
  - Website dark mode and customizable colors
    - Customizable profile picture border
    - Customizable link style
  - Editor dark mode and customizable colors
  - Scratch 2.0 -> 3.0
  - Customizable navigation bar
    - Expandable search bar
  - Customizable editor menu bar
    - Messages in editor
    - Unshare button in editor
    - Block count
    - Larger "Save Now" button
  - *Code editor*
    - Data category tweaks
    - Auto-hiding block palette
    - Two-column category menu
    - Block palette category icons
    - Customizable code area zoom
    - Editor sound effects
    - Customizable block colors
      - Customizable block shape
      - Block transparency
      - Customizable block text style
      - Square block inputs
      - Alternating nested block colors
      - Colored context menus
      - Cat blocks
  - *Costume editor*
    - *Color picker*
      - 2D color picker
      - Number inputs in color picker
  - *Stage & sprites*
    - Enhanced fullscreen
    - Hide stage and sprite pane
    - Collapsing sprite properties
    - Display stage on left side
    - Reverse order of project controls
    - Remove curved stage border
    - Progress bar
    - Clone counter
    - Mouse position display
  - *Backpack*
    - Multiple rows in backpack
    - Replace backpack bar with button
  - *Profile pages*
    - Live featured project
    - Profile page banner 
    - Show user IDs
  - *Project pages*
    - Project notes tabs
    - Sprite and script count
    - Shared/edited dates tooltip
    - Hide project stats
  - *Comments*
    - Better emojis
    - Highlight project creator in comments
    - Resizable comment input
    - Line breaks in comments
  - *Forums*
    - Better forum post editor buttons
    - 3.0 Scratchblocks on forums
    - Customizable quotes and code blocks on forums
    - Hide forum signatures
    - Necropost highlighter
    - Post cooldown display
  - *Layout*
    - Old studio layout
    - Compact editor
    - Compact messages
    - Horizontal tabs on My Stuff
    - More items per row
  - *Other*
    - Show exact count
    - High resolution thumbnails
    - More links
    - Front page curator link
    - Infinite scrolling
    - Sticky footer
    - Sound duration display
    - Hide delete button
  - *Extension popup*
    - Message count on extension icon
    - Scratch Messaging
    - Cloud games
- *Control*
  - Pause button
  - TurboWarp button
  - Higher project framerate mode
  - Debugger
  - Project video recorder
  - Gamepad support
  - Project volume slider
    - Muted project player mode
  - Running block border
  - Faster project loading
- *Community*
  - *Project pages*
    - Thumbnails setter
    - Project page download button
    - Preview project instructions and notes
    - Remix tree button
    - Remix button in own projects
  - Studio tools
    - Browse followers button
  - Full areas
  - Message filters
  - *Comments*
    - Emoji picker for comments
    - Copy link to comment button
  - Feature unshared projects
- *Forums*
  - ocular integration
  - More toolbar buttons
  - Live post preview
  - *Quoting*
    - Quote post number
    - Better quoter
  - Follow topics after posting
  - BBCode source button
- *Search*
  - Insert blocks by name
  - Editor find bar
  - Block dropdown search
  - Search box in sprite pane
  - Search bar on My Stuff
  - Username search
- *Shortcuts*
  - Ctrl+Enter to post
  - Grab single block with Ctrl key
  - Duplicate script with Alt key
  - Pick colors on stage with eyedropper
  - Arrow key incrementation in editor
  - Move costume to top or bottom
  - Move sprite to front layer
  - One-click studio invites
  - Copy code button
  - Share through My Stuff
  - Always show number pad
  - *Defaults*
    - Higher character limit in "What I'm Working On"
    - Customizable default project
    - Automatically add extensions
    - Paint costume by default
    - Customizable default costume editor colors
    - Customizable new sprite position
    - Hide new variables
    - True forums YouTube links
    - YouTube full screen
    - Redirect to standard forums
    - Remember collapsed forum categories
    - Remove autocomplete from search bar
- *Account*
  - Scratch Notifier
  - Confirmations
    - Sprite deletion confirmation
  - Disable auto-save

</details>

**Comment by DNin01** [comment]
It might be nice to have a way to narrow things down to just recommended or recommended+featured addons for an easier start. This could be a filter option or a dedicated section (like Enabled).

It recently came to my realization that it can be helpful to have recommendations above everything else (which is part of our current design) - you don't have to scroll past bunches of addons that are less likely to interest you.

That could also be remediated by enabling many of the addons by default, which is fine as long as they are intuitive.

Another way we can improve is by changing how the addons are presented, and adding shorter descriptions (or summaries) when exploring addons so it doesn't take as long to see if you want to use an addon.

**Comment by WorldLanguages** [comment]
I have said this in the Discord server already, but I don't think I did here on GitHub: **the settings page should be a separate project with its own maintainers, issues, PRs, and public APIs.** It's just one page, but it's an enormous UI/UX challenge that deserves specific dedication instead of getting covered by other pull requests.

The fact the settings page is so coupled with the rest of the extension has created multiple problems:
- No innovation and hard to make changes. (Basically, this is what we solve for Scratch thanks to the extension capability in browsers).
- No customization for the built-in settings page, with the only workaround being downloading from source, because an extension can't affect another extension.
- Also related to the item above, no competition. For example @iqnite could implement his idea by manually tagging each addon and building a separate settings page. But he won't do so because, surely, it won't get merged, and no one will use it.
- Other projects that could take advantage of the settings page UI either build their own or have trouble porting it (TurboWarp and SA-Userscript respectively)

This is what I think we should attempt for the long term then:
- The extension should expose an API to the settings page instead of the settings page calling `chrome.*` methods directly.
- Localization in the settings page should be handled with our own localization system instead of `chrome.i18n` for more flexibility, including the ability to choose any language for the settings page.
- At this point it wouldn't matter if the default settings page is Vue, React, Svelte or whatever, it's up to the maintainers of it.
- This might be the most important point. We can allow alternate setting pages (developed by anyone else) in Scratch Addons as long as they're reviewed for security and privacy. Users can then, if they want, choose which settings page to use. If an alternate settings page is universally agreed to be better in every aspect, it can become the new default and existing users can still rollback to the old one if they want.

**Reply by DNin01** [reply]
I think it would also be cool if the addon APIs were under a more permissive license such as MIT or MPL-2.0.

**Reply by WorldLanguages** [reply]
Yeah good point. If we go ahead with this, the setting pages should be under MIT License.

**Reply by iqnite** [reply]
Very good point, after all, Windows' Settings are in a separate app too.

**Reply by WorldLanguages** [reply]
Clarification: when I mention the settings page should be a separate project, I don't only mean the UI. It could also be anything related to settings management such as storing recent changes to settings, "superpresets", etc.

**Reply by DNin01** [reply]
> > I think it would also be cool if the addon APIs were under a more permissive license such as MIT or MPL-2.0.
>
> Yeah good point. If we go ahead with this, the setting pages should be under MIT License.

As future me, let me ask you, do you want competition to be able to use our open-source code in proprietary (closed-source) products that they then e.g. sell for money?

See also https://github.com/ScratchAddons/ScratchAddons/discussions/7712#discussioncomment-11825935

**Comment by DNin01** [comment]
Related https://github.com/ScratchAddons/ScratchAddons/issues/6329#issuecomment-2461193841:

> Like W_L said in https://github.com/ScratchAddons/ScratchAddons/discussions/7712#discussioncomment-10279895, since we're an open-source platform, basically, I think making Scratch Addons more modular would benefit community development. I think it could be broken up into discrete parts — UI, common libraries, and the addons themselves — that, on top of a core layer with the framework and APIs needed for addons to work, come together to make Scratch Addons as a whole, and parts can be swapped out as desired, whether you want non-official addons or a non-official UI to manage your addons.
>
> It's kind of like how apps run on operating systems which use kernels which run on hardware.

**Comment by DNin01** [comment]
On the topic of changing licenses...

All of the addons we've made are under a GPL license (and if one thing is GPL, everything becomes GPL, so Scratch Addons as a whole is going to have to be a GPL-licensed project), so we can't easily (and probably shouldn't) change that, but we might be able to change some of the other pieces of Scratch Addons.

If we decoupled the core components of Scratch Addons like the addon APIs into a standalone library, those wouldn't have to be GPL-licensed (if we rewrote them) and we could still use them in Scratch Addons by importing them as npm modules or something.

Separating those core components would also make it easier for other people to make their own Scratch browser extensions with different toolboxes, using our components sort of as their platform.

I think the underlying components and frameworks that make our addons work are begging the most to be decoupled. I've been thinking about this modular future a lot in the past couple months.

**Reply by CST1229** [reply]
> If we decoupled the core components of Scratch Addons like the addon APIs into a standalone library, those wouldn't have to be GPL-licensed (if we rewrote them)

or if we get permission from all contributors who contributed to that code i think?

**Reply by WorldLanguages** [reply]
> I think the underlying components and frameworks that make our addons work are begging the most to be decoupled.

Components such as...?

**Reply by DNin01** [reply]
> > I think the underlying components and frameworks that make our addons work are begging the most to be decoupled.
> 
> Components such as...?

The extension's general inner workings (a.k.a. core), addon APIs/libraries, etc.

**Comment by DNin01** [comment]
### United code is a mess

At this point, because everything in the extension is so coupled together, it's also disorganized. You can't always tell what is used by what, and this makes our code hard to take apart. If you want to use our underlying code, you'll most likely just end up with a reskin of Scratch Addons or rebuilding everything from the ground up.

All this does is remind me of the proprietary software model.

### If we could start over...

But if we could untie the knots, I think a few things would change:
- The addons, the core and APIs, and the UI of Scratch Addons would be divided. They could either all be in separate repositories or all in the same repository in different source-code-top-level directories, but they _would_ all each be separate development projects.
- Each of these segments would also have their own libraries, so if someone decides to rebuild one of them from the ground up, such as all of the addons, they can remove all the libraries associated with them.
- Wherever it makes sense and wherever we can, we should leave room for expansion, so if for some reason a developer wanted to add a whole new capability to Scratch Addons, it wouldn't be too hard to connect the dots.

The advantage of using the structure/system I just described is that forking would be super simple. If you wanted to swap out all the addons or reskin the extension or add new capabilities, there's one place to do each of those.

### What about competition?

This would open the floodgates for competition, but for FOSS, that's only a good thing, and there are things we can do to keep it that way, like for example, if we want to make sure the competition doesn't take our extension and fork it into something proprietary, it should all be GPL-licensed.

**Reply by WorldLanguages** [reply]
I agree with your points and motivation, but can't say a lot unless they become more concrete suggestions.
By the way, Scratch's browser extension policy distorts competition in the space and that's not something we control.

**Reply by WorldLanguages** [reply]
> if we want to make sure the competition doesn't take our extension and fork it into something proprietary, it should all be GPL-licensed.

Personally I would consider relicensing all of my code to MIT license, I don't really think any proprietary software will be built similar to ours.
If I'm ever proven wrong, I wonder if they'll train an AI with our code and argue its outputs are in the public domain 😅
